solutions/platinum/csa-RootLCAQueries.mdx

```diff
@@ -0,0 +1,138 @@
+---
+id: csa-RootLCAQueries
+source: CS Academy
+title: Root LCA Queries
+author: Justin Ji
+---
+
+<Spoiler title = "Hint 1">
+
+Let $a$, $b$, and $c$ be the three nodes given in each query.
+When will the number of possible roots be greater than zero?
+
+</Spoiler>
+
+<Spoiler title = "Answer to Hint 1">
+
+The answer for a given query is only greater than zero when $c$ is on the path
+from $a$ to $b$, as by definition the LCA must lie on the path from $a$ to $b$.
+Now try doing a bit of casework to calculate the answer.
+
+</Spoiler>
+
+<Spoiler title = "Solution">
+
+As noted in the hints, the LCA of $a$ and $b$ must lie on the path between these
+nodes. Arbitrarily root the tree, then consider the following cases:
+
+**Case 1: $c$ is not on the path from $a$ to $b$.**
+
+The answer is $0$ for this case.
+
+**Case 2: $c$ is equal to the LCA of $a$ and $b$, for our chosen root.**
+
+Any node outside of the subtree of $c$ is a suitable root. Within the subtree of
+$c$, any node whose path to node $c$ does not intersect the path from $a$ to $b$
+is a suitable root.
+
+**Case 3: $c$ is on the path from $a$ to $b$, but is not equal to the LCA of
+$a$ and $b$.**
+
+This case is handled the same way as case 2, except only nodes in the subtree
+of $c$ can be possible roots.
+
+To quickly answer queries, we use binary lifting to calculate the LCA, and
+precalculate the number of nodes in each subtree to quickly remove unsuitable
+subtrees from our answer.
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(N\log{N} + Q\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int n, q;
+	cin >> n >> q;
+	vector<vector<int>> adj(n);
+	for (int i = 1; i < n; i++) {
+		int u, v;
+		cin >> u >> v;
+		adj[--u].push_back(--v);
+		adj[v].push_back(u);
+	}
+
+	const int log2dist = 1 + (int)log2(n);
+	vector<int> dep(n);
+	vector<int> sub(n, 1);
+	vector<vector<int>> lift(log2dist, vector<int>(n));
+	function<void(int, int)> dfs = [&](int u, int p) {
+		for (int v : adj[u]) {
+			if (v == p) continue;
+			dep[v] = dep[u] + 1;
+			lift[0][v] = u;
+			for (int p = 1; p < log2dist; p++) {
+				lift[p][v] = lift[p - 1][lift[p - 1][v]];
+			}
+			dfs(v, u);
+			sub[u] += sub[v];
+		}
+	};
+	dfs(0, -1);
+
+	/** @return d-th ancestor of node u */
+	auto jump = [&](int u, int d) -> int {
+		for (int i = log2dist - 1; i >= 0; i--) {
+			if ((d >> i) & 1) { u = lift[i][u]; }
+		}
+		return u;
+	};
```

Comment:
i'll kiss you if you put the lca stuff in a class

Reply:
actually uh this pr may go dark for ages bc i was planning on sansifying the bin jumping module

Reply:
ill quickly put it in a class